sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/GreatestSqlAggregator.java

```diff
@@ -0,0 +1,116 @@
+
+import org.apache.calcite.sql.SqlAggFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.math.expr.ExprMacroTable;
+import org.apache.druid.query.aggregation.AggregatorFactory;
+import org.apache.druid.query.aggregation.DoubleMaxAggregatorFactory;
+import org.apache.druid.query.aggregation.LongMaxAggregatorFactory;
+import org.apache.druid.query.aggregation.PostAggregator;
+import org.apache.druid.query.aggregation.post.DoubleGreatestPostAggregator;
+import org.apache.druid.query.aggregation.post.LongGreatestPostAggregator;
+import org.apache.druid.segment.column.ValueType;
+
+import java.util.List;
+
+/**
+ * Calcite integration class for Greatest post aggregators of Long & Double types.
+ * It applies Max aggregators over the provided fields/expressions & combines their results via Field access post aggregators.
+ */
+public class GreatestSqlAggregator extends MultiColumnSqlAggregator
+{
+  private static final SqlAggFunction FUNCTION_INSTANCE = new GreatestSqlAggFunction();
+  private static final String NAME = "GREATEST";
+
+  @Override
+  public SqlAggFunction calciteFunction()
+  {
+    return FUNCTION_INSTANCE;
+  }
+
+  @Override
+  AggregatorFactory createAggregatorFactory(
+      ValueType valueType,
+      String prefixedName,
+      FieldInfo fieldInfo,
+      ExprMacroTable macroTable
+  )
+  {
+    final AggregatorFactory aggregatorFactory;
+    switch (valueType) {
+      case LONG:
+        aggregatorFactory = new LongMaxAggregatorFactory(prefixedName, fieldInfo.fieldName, fieldInfo.expression, macroTable);
+        break;
+      case FLOAT:
+      case DOUBLE:
+        aggregatorFactory = new DoubleMaxAggregatorFactory(prefixedName, fieldInfo.fieldName, fieldInfo.expression, macroTable);
+        break;
+      default:
+        throw new ISE("Cannot create aggregator factory for type[%s]", valueType);
+    }
+    return aggregatorFactory;
+  }
+
+  @Override
+  PostAggregator createFinalPostAggregator(
+      ValueType valueType,
+      String name,
+      List<PostAggregator> postAggregators
+  )
+  {
+    final PostAggregator finalPostAggregator;
+    switch (valueType) {
+      case LONG:
+        finalPostAggregator = new LongGreatestPostAggregator(name, postAggregators);
+        break;
+      case FLOAT:
+      case DOUBLE:
+        finalPostAggregator = new DoubleGreatestPostAggregator(name, postAggregators);
+        break;
+      default:
+        throw new ISE("Cannot create aggregator factory for type[%s]", valueType);
+    }
+    return finalPostAggregator;
+  }
+
+  private static class GreatestSqlAggFunction extends SqlAggFunction
+  {
+    GreatestSqlAggFunction()
+    {
+      super(
+          NAME,
+          null,
+          SqlKind.GREATEST,
+          ReturnTypes.ARG0_NULLABLE_IF_EMPTY,
```

Comment:
The GREATEST and LEAST SQLFunctions use this as the return type - `ReturnTypes.cascade(ReturnTypes.LEAST_RESTRICTIVE, SqlTypeTransforms.TO_NULLABLE)`, why did you choose ARG0_NULLABLE_IF_EMPTY?

Reply:
I had picked the type of the first argument but the cascade you mentioned is a much better fit.
Updated.